File: BellmanFord.java
import java.util.*;

public class BellmanFord {
    static class Edge {
        int source, destination, weight;
        
        Edge(int source, int destination, int weight) {
            this.source = source;
            this.destination = destination;
            this.weight = weight;
        }
    }
    
    private int vertices;
    private List<Edge> edges;
    
    public BellmanFord(int vertices) {
        this.vertices = vertices;
        this.edges = new ArrayList<>();
    }
    
    public void addEdge(int source, int destination, int weight) {
        edges.add(new Edge(source, destination, weight));
    }
    
    public void bellmanFord(int source) {
        int[] distance = new int[vertices];
        Arrays.fill(distance, Integer.MAX_VALUE);
        distance[source] = 0;
        
        // Relax all edges V-1 times
        for (int i = 0; i < vertices - 1; i++) {
            for (Edge edge : edges) {
                if (distance[edge.source] != Integer.MAX_VALUE && 
                    distance[edge.source] + edge.weight < distance[edge.destination]) {
                    distance[edge.destination] = distance[edge.source] + edge.weight;
                }
            }
        }
        
        // Check for negative cycles
        for (Edge edge : edges) {
            if (distance[edge.source] != Integer.MAX_VALUE && 
                distance[edge.source] + edge.weight < distance[edge.destination]) {
                System.out.println("Graph contains negative cycle");
                return;
            }
        }
        
        // Print distances
        for (int i = 0; i < vertices; i++) {
            System.out.println("Distance from " + source + " to " + i + ": " + distance[i]);
        }
    }
    
    public static void main(String[] args) {
        BellmanFord graph = new BellmanFord(5);
        graph.addEdge(0, 1, -1);
        graph.addEdge(0, 2, 4);
        graph.addEdge(1, 2, 3);
        graph.addEdge(1, 3, 2);
        graph.addEdge(1, 4, 2);
        graph.addEdge(3, 2, 5);
        graph.addEdge(3, 1, 1);
        graph.addEdge(4, 3, -3);
        
        graph.bellmanFord(0);
    }
}

File: Python/graph/bellman_ford.py
class BellmanFord:
    def __init__(self, vertices):
        self.vertices = vertices
        self.edges = []
    
    def add_edge(self, source, destination, weight):
        self.edges.append((source, destination, weight))
    
    def bellman_ford(self, source):
        distance = [float('inf')] * self.vertices
        distance[source] = 0
        
        # Relax all edges V-1 times
        for _ in range(self.vertices - 1):
            for source_v, dest_v, weight in self.edges:
                if distance[source_v] != float('inf') and distance[source_v] + weight < distance[dest_v]:
                    distance[dest_v] = distance[source_v] + weight
        
        # Check for negative cycles
        for source_v, dest_v, weight in self.edges:
            if distance[source_v] != float('inf') and distance[source_v] + weight < distance[dest_v]:
                print("Graph contains negative cycle")
                return
        
        # Print distances
        for i in range(self.vertices):
            print(f"Distance from {source} to {i}: {distance[i]}")

if __name__ == "__main__":
    graph = BellmanFord(5)
    graph.add_edge(0, 1, -1)
    graph.add_edge(0, 2, 4)
    graph.add_edge(1, 2, 3)
    graph.add_edge(1, 3, 2)
    graph.add_edge(1, 4, 2)
    graph.add_edge(3, 2, 5)
    graph.add_edge(3, 1, 1)
    graph.add_edge(4, 3, -3)
    
    graph.bellman_ford(0)